File: blink-macos/Blink/UpdateChecker.swift
import AppKit

/// Checks GitHub Releases for a newer version on launch.
@MainActor
final class UpdateChecker: ObservableObject {
    static let shared = UpdateChecker()

    enum CheckResult: Equatable {
        case upToDate
        case available(String)
        case failed
    }

    @Published var updateAvailable: Bool = false
    @Published var latestVersion: String?
    @Published var downloadURL: URL?
    @Published var isChecking: Bool = false
    @Published var lastCheckResult: CheckResult?
    private static let currentVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
    private static let releasesURL = URL(string: "https://api.github.com/repos/D4G4/blink/releases/latest")!

    /// True when installed from the Mac App Store (has a receipt file).
    static let isAppStore: Bool = {
        Bundle.main.appStoreReceiptURL?.path().contains("sandboxReceipt") == true ||
        Bundle.main.appStoreReceiptURL.flatMap({ FileManager.default.fileExists(atPath: $0.path()) }) == true
    }()

    enum InstallSource {
        case appStore   // sandboxed install — update checks skipped entirely
        case homebrew   // installed via the D4G4/homebrew-blink cask
        case dmg        // direct download from a GitHub release
    }

    /// Best-effort install-source detection so the update HUD can offer
    /// the right upgrade path. Homebrew users get a `brew upgrade` command
    /// to copy; DMG users get a direct download link.
    ///
    /// Detection heuristic for Homebrew: check whether `Caskroom/blink/`
    /// contains a versioned subdirectory (e.g. `1.4.0/`) under either
    /// standard prefix (`/opt/homebrew` for Apple Silicon, `/usr/local`
    /// for Intel). Checking only for the parent dir's existence
    /// mis-detects users who once installed via brew and later uninstalled
    /// — brew sometimes leaves an empty `Caskroom/blink/` behind. A
    /// non-empty subdirectory listing is the reliable signal that brew
    /// currently tracks an install.
    static let installSource: InstallSource = {
        if isAppStore { return .appStore }
        let fm = FileManager.default
        for cask in ["/opt/homebrew/Caskroom/blink", "/usr/local/Caskroom/blink"] {
            if let contents = try? fm.contentsOfDirectory(atPath: cask),
               !contents.filter({ !$0.hasPrefix(".") }).isEmpty {
                return .homebrew
            }
        }
        return .dmg
    }()

    private var periodicTimer: Timer?

    func startPeriodicChecks() {
        guard !Self.isAppStore else {
            Log.i("App Store install — skipping update checks")
            return
        }
        checkForUpdate()
        periodicTimer = Timer.scheduledTimer(withTimeInterval: 86400, repeats: true) { [weak self] _ in
            MainActor.assumeIsolated {
                self?.checkForUpdate()
            }
        }
    }

    func checkForUpdate() {
        isChecking = true
        lastCheckResult = nil
        Task {
            do {
                var request = URLRequest(url: Self.releasesURL)
                request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
                request.timeoutInterval = 10

                let (data, _) = try await URLSession.shared.data(for: request)

                guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                      let tagName = json["tag_name"] as? String else {
                    isChecking = false
                    lastCheckResult = .failed
                    return
                }

                let latest = tagName.replacingOccurrences(of: "v", with: "")
                let current = Self.currentVersion

                if isNewer(latest, than: current) {
                    Log.i("Update available: \(current) → \(latest)")
                    latestVersion = latest
                    updateAvailable = true
                    lastCheckResult = .available(latest)

                    // Find DMG download URL from assets
                    if let assets = json["assets"] as? [[String: Any]] {
                        for asset in assets {
                            if let name = asset["name"] as? String, name.hasSuffix(".dmg"),
                               let url = asset["browser_download_url"] as? String {
                                downloadURL = URL(string: url)
                                break
                            }
                        }
                    }

                    // Fallback to release page
                    if downloadURL == nil, let htmlURL = json["html_url"] as? String {
                        downloadURL = URL(string: htmlURL)
                    }
                } else {
                    Log.i("Up to date: \(current)")
                    lastCheckResult = .upToDate
                }
            } catch {
                Log.d("Update check failed: \(error.localizedDescription)")
                lastCheckResult = .failed
            }
            isChecking = false
        }
    }

    static let brewCommand = "brew update && brew upgrade --cask blink"

    /// Simple semver comparison: "1.2.0" > "1.1.0"
    private func isNewer(_ a: String, than b: String) -> Bool {
        let aParts = a.split(separator: ".").compactMap { Int($0) }
        let bParts = b.split(separator: ".").compactMap { Int($0) }

        for i in 0..<max(aParts.count, bParts.count) {
            let av = i < aParts.count ? aParts[i] : 0
            let bv = i < bParts.count ? bParts[i] : 0
            if av > bv { return true }
            if av < bv { return false }
        }
        return false
    }
}

File: blink-windows/src/Blink.App/UpdateChecker.cs
using System.ComponentModel;
using System.Net.Http;
using System.Runtime.CompilerServices;
using System.Text.Json;

namespace Blink.App;

/// <summary>
/// Checks GitHub releases for newer versions.
/// Port of macOS UpdateChecker.swift.
/// </summary>
public sealed class UpdateChecker : INotifyPropertyChanged
{
    public static UpdateChecker Instance { get; } = new();

    // -- CheckResult ----------------------------------------------------------

    public enum CheckResultKind { UpToDate, Available, Failed }

    public sealed record CheckResult(CheckResultKind Kind, string? AvailableVersion = null)
    {
        public static CheckResult UpToDate => new(CheckResultKind.UpToDate);
        public static CheckResult Available(string version) => new(CheckResultKind.Available, version);
        public static CheckResult Failed => new(CheckResultKind.Failed);
    }

    // -- Observable properties ------------------------------------------------

    private bool _updateAvailable;
    public bool UpdateAvailable
    {
        get => _updateAvailable;
        private set => Set(ref _updateAvailable, value);
    }

    private string? _latestVersion;
    public string? LatestVersion
    {
        get => _latestVersion;
        private set => Set(ref _latestVersion, value);
    }

    private string? _downloadUrl;
    public string? DownloadUrl
    {
        get => _downloadUrl;
        private set => Set(ref _downloadUrl, value);
    }

    private bool _isChecking;
    public bool IsChecking
    {
        get => _isChecking;
        private set => Set(ref _isChecking, value);
    }

    private CheckResult? _lastCheckResult;
    public CheckResult? LastCheckResult
    {
        get => _lastCheckResult;
        private set => Set(ref _lastCheckResult, value);
    }

    // -- Internals ------------------------------------------------------------

    private static readonly string CurrentVersion =
        typeof(UpdateChecker).Assembly.GetName().Version?.ToString(3) ?? "1.0.0";

    private const string ReleasesUrl =
        "https://api.github.com/repos/D4G4/blink/releases/latest";

    private static readonly HttpClient Http = new()
    {
        Timeout = TimeSpan.FromSeconds(10),
        DefaultRequestHeaders =
        {
            { "Accept", "application/vnd.github+json" },
            { "User-Agent", "Blink-Windows-UpdateChecker" }
        }
    };

    private System.Threading.Timer? _periodicTimer;

    // -- Public API -----------------------------------------------------------

    public void StartPeriodicChecks()
    {
        CheckForUpdate();
        _periodicTimer = new System.Threading.Timer(
            _ => CheckForUpdate(),
            null,
            TimeSpan.FromMilliseconds(86_400_000),
            TimeSpan.FromMilliseconds(86_400_000));
    }

    public async void CheckForUpdate()
    {
        IsChecking = true;
        LastCheckResult = null;

        try
        {
            var json = await Http.GetStringAsync(ReleasesUrl);
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            if (!root.TryGetProperty("tag_name", out var tagProp))
            {
                IsChecking = false;
                LastCheckResult = CheckResult.Failed;
                return;
            }

            var latest = tagProp.GetString()?.Replace("v", "") ?? "";
            var current = CurrentVersion;

            if (IsNewer(latest, current))
            {
                LatestVersion = latest;
                UpdateAvailable = true;
                LastCheckResult = CheckResult.Available(latest);

                // Look for .exe or .msi asset
                if (root.TryGetProperty("assets", out var assets))
                {
                    foreach (var asset in assets.EnumerateArray())
                    {
                        if (asset.TryGetProperty("name", out var nameProp))
                        {
                            var name = nameProp.GetString() ?? "";
                            if (name.EndsWith(".exe") || name.EndsWith(".msi"))
                            {
                                if (asset.TryGetProperty("browser_download_url", out var urlProp))
                                {
                                    DownloadUrl = urlProp.GetString();
                                    break;
                                }
                            }
                        }
                    }
                }

                // Fallback to release page
                if (DownloadUrl == null && root.TryGetProperty("html_url", out var htmlUrl))
                {
                    DownloadUrl = htmlUrl.GetString();
                }
            }
            else
            {
                LastCheckResult = CheckResult.UpToDate;
            }
        }
        catch
        {
            LastCheckResult = CheckResult.Failed;
        }

        IsChecking = false;
    }

    // -- Semver comparison ----------------------------------------------------

    internal static bool IsNewer(string a, string b)
    {
        var aParts = a.Split('.').Select(s => int.TryParse(s, out var n) ? n : 0).ToArray();
        var bParts = b.Split('.').Select(s => int.TryParse(s, out var n) ? n : 0).ToArray();

        var len = Math.Max(aParts.Length, bParts.Length);
        for (var i = 0; i < len; i++)
        {
            var av = i < aParts.Length ? aParts[i] : 0;
            var bv = i < bParts.Length ? bParts[i] : 0;
            if (av > bv) return true;
            if (av < bv) return false;
        }
        return false;
    }

    // -- INotifyPropertyChanged -----------------------------------------------

    public event PropertyChangedEventHandler? PropertyChanged;

    private void Set<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value)) return;
        field = value;
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
    }
}
